### PR TITLE
Promisify public API

### PR DESCRIPTION
**Pull request by s1na**
I'd appreciate having a broader look here to see I haven't missed anything.

As per discussion in #455 and #544

### CODE REVIEW

**Comment by holgerd77** [comment]
There is a WASM file in here. ? 😋

**Comment by s1na** [comment]
Oops :smile: Removed and rebased

File: tests/api/runCode.js [hunk 1]
```diff
@@ -15,45 +15,35 @@ const testCases = [
...
+        err = err ? err.message : 'no error thrown'
+        st.equals(err, testData.error, 'error message should match')
+        err = false
+      }
 
-        t.assert(!err)
-        st.end()
-      })
+      st.assert(!err)
+      st.end()
```
Comment (holgerd77) [inline tests/api/runCode.js]: Post-review comment: this is also a nice simplification.

File: lib/index.ts [hunk 2]
```diff
@@ -111,39 +112,38 @@ export default class VM extends AsyncEventEmitter {
    * @param blockchain -  A [blockchain](https://github.com/ethereum/ethereumjs-blockchain) object to process
    * @param cb - the callback function
    */
-  runBlockchain(blockchain: any, cb: any): void {
-    runBlockchain.bind(this)(blockchain, cb)
+  runBlockchain(blockchain: any): Promise<void> {
+    return promisify(runBlockchain.bind(this))(blockchain)
   }
```
[the remark this answers is not in the document]
Comment (holgerd77) [reply lib/index.ts]: Update: explanation [here](https://github.com/ethereumjs/ethereumjs-vm/pull/517#issue-281514326), no need to give an answer.

[the remark this answers is not in the document]
Comment (s1na) [reply lib/index.ts]: > explanation here, no need to give an answer.

Yeah that was the reason, but on the other hand what I could have done is to return a promise in `runBlockchain` without using promises within it internally. I did this now and pushed, so we won't need to `promisify` in the VM.

File: lib/runTx.ts [hunk 3]
```diff
@@ -80,23 +68,21 @@ export default function runTx(this: VM, opts: RunTxOpts, cb: RunTxCb) {
...
+    const result = await _runTx.bind(this)(opts)
+    await state.commit()
+    return result
+  } catch (e) {
+    await state.revert()
+    throw e
+  }
 }
 
 async function _runTx(this: VM, opts: RunTxOpts): Promise<RunTxResult> {
```
[the remark this answers is not in the document]
Comment (s1na) [reply lib/runTx.ts]: I thought it's a clean separation that the outer function performs checkpointing, commit on success and revert on any error. And the internal method modifies the state without being aware of this.

I can change this however if it doesn't make sense

[the remark this answers is not in the document]
Comment (holgerd77) [reply lib/runTx.ts]: Ah, no, that's a good reasoning, thanks.

**Pull request merged**